File: front/lib/api/src/models/chapter.dart
import 'package:blee/api/src/models/image.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'chapter.freezed.dart';
part 'chapter.g.dart';

@freezed
class Chapter with _$Chapter {
  const factory Chapter({
    required String id,
    required String name,
    @JsonKey(name: "movie_id")
    required String movieId,
    @JsonKey(name: "start_time")
    required int startTime,
    @JsonKey(name: "end_time")
    required int endTime,
    required Image? thumbnail,
    required List<ChapterType> type,
  }) = _Chapter;

  factory Chapter.fromJson(Map<String, dynamic> json) => _$ChapterFromJson(json);
}

enum ChapterType {
  @JsonValue("interview")
  Interview,
  @JsonValue("non_musical_interview")
  NonMusicalInterview,
  @JsonValue("other")
  Other,
  @JsonValue("performance")
  Performance
}

File: front/models/domain/chapter.ts
import * as yup from "yup";
import { Resource } from "./resource";
import { Image } from "./image";

const ChapterTypes = [
    "interview",
    "non_musical_interview",
    "other",
    "performance",
];

type ChapterType = (typeof ChapterTypes)[number];

const Chapter = Resource.concat(
    yup.object({
        name: yup.string().required(),
        movie_id: yup.string().required(),
        start_time: yup.number().required(),
        end_time: yup.number().required(),
        thumbnail: Image.required().nullable(),
        type: yup
            .array(yup.mixed<ChapterType>().oneOf(ChapterTypes).required())
            .required(),
    }),
);

type Chapter = yup.InferType<typeof Chapter>;

export { type ChapterType, Chapter };
